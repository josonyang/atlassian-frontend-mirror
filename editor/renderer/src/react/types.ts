import type { PropsWithChildren } from 'react';
import type { Node as PMNode, NodeType, Schema } from '@atlaskit/editor-prosemirror/model';
import type { ExtensionHandlers } from '@atlaskit/editor-common/extensions';
import type { ProviderFactory } from '@atlaskit/editor-common/provider-factory';
import type { EventHandlers } from '@atlaskit/editor-common/ui';
import type { AnalyticsEventPayload } from '../analytics/events';
import type { Serializer } from '../serializer';
import type { RendererAppearance, HeadingAnchorLinksProps } from '../ui/Renderer/types';
import type { AnnotationId, AnnotationTypes } from '@atlaskit/adf-schema';

export interface RendererContext {
	objectAri?: string;
	containerAri?: string;
	// Ignored via go/ees005
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	adDoc?: any;
	schema?: Schema;
}

export interface NodeMeta {
	text?: PMNode['text'];
	providers?: ProviderFactory | undefined;
	eventHandlers?: EventHandlers | undefined;
	extensionHandlers?: ExtensionHandlers | undefined;
	portal?: HTMLElement | undefined;
	rendererContext?: RendererContext;
	serializer: Serializer<JSX.Element>;
	content?: {
		// Ignored via go/ees005
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		[key: string]: any;
	} | null;
	allowHeadingAnchorLinks?: HeadingAnchorLinksProps;
	allowCopyToClipboard?: boolean;
	allowWrapCodeBlock?: boolean;
	allowPlaceholderText?: boolean;
	allowCustomPanels?: boolean;
	rendererAppearance?: RendererAppearance;
	fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
	nodeType: NodeType['name'];
	marks: PMNode['marks'];
	dataAttributes: {
		'data-renderer-start-pos': number;
	};
	// Ignored via go/ees005
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[key: string]: any;
}

export interface MarkMeta {
	dataAttributes: {
		'data-renderer-mark': true;
		'data-block-mark'?: true;
	};
	eventHandlers?: EventHandlers;
	fireAnalyticsEvent?: (event: AnalyticsEventPayload) => void;
	// Ignored via go/ees005
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	markKey?: any;
	// Whether the node this mark belongs to is an inline node, if available
	isInline?: boolean;
	// Ignored via go/ees005
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	[key: string]: any;
}

export interface AnnotationMarkMeta extends MarkMeta {
	id: AnnotationId;
	annotationType: AnnotationTypes;
	annotationParentIds: string[];
	allowAnnotations: boolean;
	useBlockLevel?: boolean;
	isMediaInline?: boolean;
}

export type NodeProps<NodeAttrs = Object> = NodeAttrs & PropsWithChildren<NodeMeta>;
export type MarkProps<MarkAttrs = Object> = MarkAttrs & PropsWithChildren<MarkMeta>;

export type TextHighlighter = {
	pattern: RegExp;
	component: React.ComponentType<{
		children: React.ReactNode;
		match: string;
		marks: Set<string>; // In future maybe extract the mark names from Schema
		groups: Array<string> | undefined;
		startPos: number;
	}>;
};
