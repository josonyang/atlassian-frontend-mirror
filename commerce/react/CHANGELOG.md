# @af/commerce-react

## 0.1.3

### Patch Changes

- [`61cb5313358`](https://bitbucket.org/atlassian/atlassian-frontend/commits/61cb5313358) - Removing unused dependencies and dev dependencies

## 0.1.2

### Patch Changes

- [`9d00501a414`](https://bitbucket.org/atlassian/atlassian-frontend/commits/9d00501a414) - Ensure legacy types are published for TS 4.5-4.8

## 0.1.1

### Patch Changes

- [`41fae2c6f68`](https://bitbucket.org/atlassian/atlassian-frontend/commits/41fae2c6f68) - Upgrade Typescript from `4.5.5` to `4.9.5`

## 0.1.0

### Minor Changes

- [`56507598609`](https://bitbucket.org/atlassian/atlassian-frontend/commits/56507598609) - Skip minor dependency bump

## 0.0.5

### Patch Changes

- [`5806023831a`](https://bitbucket.org/atlassian/atlassian-frontend/commits/5806023831a) - Internal change to enforce token usage for spacing properties. There is no expected visual or behaviour change.

## 0.0.4

### Patch Changes

- [`8cc2f888c83`](https://bitbucket.org/atlassian/atlassian-frontend/commits/8cc2f888c83) - Upgrade Typescript from `4.3.5` to `4.5.5`

## 0.0.3

### Patch Changes

- [`8d4228767b0`](https://bitbucket.org/atlassian/atlassian-frontend/commits/8d4228767b0) - Upgrade Typescript from `4.2.4` to `4.3.5`.

## 0.0.2

### Patch Changes

- [`cb2392f6d33`](https://bitbucket.org/atlassian/atlassian-frontend/commits/cb2392f6d33) - Upgrade to TypeScript 4.2.4

## 0.0.1

### Patch Changes

- [`e5c858ab40c`](https://bitbucket.org/atlassian/atlassian-frontend/commits/e5c858ab40c) - Updating stripe and managing state update

## 0.0.4

### Patch Changes

- [`229b32842b5`](https://bitbucket.org/atlassian/atlassian-frontend/commits/229b32842b5) - Fix .npmignore and tsconfig.json for **tests**

## 0.0.3

### Patch Changes

- [`d3265f19be`](https://bitbucket.org/atlassian/atlassian-frontend/commits/d3265f19be) - Transpile packages using babel rather than tsc

## 0.0.2

### Patch Changes

- [`5f58283e1f`](https://bitbucket.org/atlassian/atlassian-frontend/commits/5f58283e1f) - Export types using Typescript's new "export type" syntax to satisfy Typescript's --isolatedModules compiler option.
  This requires version 3.8 of Typescript, read more about how we handle Typescript versions here: https://atlaskit.atlassian.com/get-started
  Also add `typescript` to `devDependencies` to denote version that the package was built with.

## 0.0.1

### Patch Changes

- [`b443b5a60f`](https://bitbucket.org/atlassian/atlassian-frontend/commits/b443b5a60f) - Renamed template package
