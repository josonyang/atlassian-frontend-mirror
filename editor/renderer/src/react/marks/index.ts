import type { ComponentType } from 'react';
import type { Mark } from '@atlaskit/editor-prosemirror/model';
import Code from './code';
import Em from './em';
import Link from './link';
import Strike from './strike';
import Strong from './strong';
import Subsup from './subsup';
import TextColor from './textColor';
import BackgroundColor from './backgroundColor';
import Underline from './underline';
import Breakout from './breakout';
import Alignment from './alignment';
import Indentation from './indentation';
import UnsupportedMark from './unsupportedMark';
import UnsupportedNodeAttribute from './unsupportedNodeAttribute';
import DataConsumer from './data-consumer';
import FragmentMark from './fragment';
import Annotation, { isAnnotationMark } from './annotation';
import Border from './border';

// Stage0
import ConfluenceInlineComment from './confluence-inline-comment';

export const markToReact: {
	[key: string]: ComponentType<React.PropsWithChildren<any>>;
} = {
	code: Code,
	em: Em,
	link: Link,
	strike: Strike,
	strong: Strong,
	subsup: Subsup,
	textColor: TextColor,
	backgroundColor: BackgroundColor,
	underline: Underline,
	annotation: Annotation,
	border: Border,

	// Stage0
	confluenceInlineComment: ConfluenceInlineComment,
	breakout: Breakout,
	alignment: Alignment,
	indentation: Indentation,
	unsupportedMark: UnsupportedMark,
	unsupportedNodeAttribute: UnsupportedNodeAttribute,
	dataConsumer: DataConsumer,
	fragment: FragmentMark,
};

export const toReact = (mark: Mark): ComponentType<React.PropsWithChildren<any>> => {
	return markToReact[mark.type.name];
};

export {
	Code,
	Em,
	Link,
	Strike,
	Strong,
	Subsup,
	TextColor,
	BackgroundColor,
	Underline,
	Breakout,
	Annotation,
	Border,
	UnsupportedMark,
	isAnnotationMark,
	UnsupportedNodeAttribute,
	DataConsumer,
	FragmentMark,
};
