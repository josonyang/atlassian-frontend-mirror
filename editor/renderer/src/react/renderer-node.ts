import type { Node, Schema } from '@atlaskit/editor-prosemirror/model';

export const isLayoutNode = (node: Node) => node.type.name === 'layoutSection';

export const hasBreakOutMark = (node: Node) =>
  node.marks.some((m) => m.type.name === 'breakout');

export const insideBreakoutLayout = (path: Node[]) =>
  path.some((item) => isLayoutNode(item) && hasBreakOutMark(item));

export const insideBlockNode = (path: Node[], schema: Schema) => {
  const {
    nodes: { expand, nestedExpand, layoutColumn },
  } = schema;
  const blockNodeNames = [expand, nestedExpand, layoutColumn]
    .filter((node) => Boolean(node))
    .map((node) => node.name);

  return (
    path && path.some((n) => n.type && blockNodeNames.indexOf(n.type.name) > -1)
  );
};

export const insideMultiBodiedExtension = (path: Node[], schema: Schema) => {
  const {
    nodes: { multiBodiedExtension },
  } = schema;
  return path.some((n) => n.type === multiBodiedExtension);
};
