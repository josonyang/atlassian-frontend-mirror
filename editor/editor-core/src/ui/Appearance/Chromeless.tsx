/**
 * @jsxRuntime classic
 * @jsx jsx
 */
import React, { Fragment } from 'react';

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766
import { css, jsx } from '@emotion/react';

import { useSharedPluginStateWithSelector } from '@atlaskit/editor-common/hooks';
import type { EditorAppearance, OptionalPlugin } from '@atlaskit/editor-common/types';
import type { EditorViewModePlugin } from '@atlaskit/editor-plugins/editor-viewmode';
import type {
	MaxContentSizePlugin,
	MaxContentSizePluginState,
} from '@atlaskit/editor-plugins/max-content-size';
import { fg } from '@atlaskit/platform-feature-flags';
import { token } from '@atlaskit/tokens';

import type { EditorAppearanceComponentProps } from '../../types';
import EditorContentContainer from '../EditorContentContainer/EditorContentContainer';
import PluginSlot from '../PluginSlot';
import WithFlash from '../WithFlash';

const scrollbarStylesNew = css({
	msOverflowStyle: '-ms-autohiding-scrollbar',
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
	'&::-webkit-scrollbar-corner': {
		display: 'none',
	},
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
	'&::-webkit-scrollbar-thumb': {
		backgroundColor: token('color.background.neutral.subtle'),
	},
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
	'&:hover::-webkit-scrollbar-thumb': {
		backgroundColor: token('color.background.neutral.bold'),
		borderRadius: token('radius.large'),
	},
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-unsafe-selectors
	'&::-webkit-scrollbar-thumb:hover': {
		backgroundColor: token('color.background.neutral.bold.hovered'),
	},
});

const chromelessEditorStylesNew = css({
	// eslint-disable-next-line @atlaskit/design-system/use-tokens-typography
	lineHeight: '20px',
	height: 'auto',
	overflowX: 'hidden',
	overflowY: 'auto',
	maxWidth: 'inherit',
	boxSizing: 'border-box',
	wordWrap: 'break-word',
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
	'div > .ProseMirror': {
		outline: 'none',
		whiteSpace: 'pre-wrap',
		padding: 0,
		margin: 0,
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
		'& > :last-child': {
			paddingBottom: token('space.100', '0.5em'),
		},
	},
});

const extraSpaceLastLineFix = css({
	// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors -- Ignored via go/DSP-18766
	'.ProseMirror': {
		// eslint-disable-next-line @atlaskit/ui-styling-standard/no-nested-selectors, @atlaskit/ui-styling-standard/no-unsafe-selectors -- Ignored via go/DSP-18766
		'& > p:last-of-type': {
			marginBottom: token('space.0'),
		},
	},
});

type AppearanceProps = EditorAppearanceComponentProps<
	[OptionalPlugin<MaxContentSizePlugin>, OptionalPlugin<EditorViewModePlugin>]
>;

/**
 * Render the editor in a chromeless appearance.
 * Example use is the inline comment editor, which doesn't have editor toolbar
 */
// Ignored via go/ees005
// eslint-disable-next-line @repo/internal/react/no-class-components
export default class Editor extends React.Component<AppearanceProps> {
	static displayName = 'ChromelessEditorAppearance';

	private appearance: EditorAppearance = 'chromeless';
	private containerElement: HTMLElement | null = null;

	private renderChrome = ({ maxContentSize }: { maxContentSize?: MaxContentSizePluginState }) => {
		const {
			editorDOMElement,
			editorView,
			editorActions,
			eventDispatcher,
			providerFactory,
			contentComponents,
			customContentComponents,
			maxHeight,
			minHeight = 30,
			popupsMountPoint,
			popupsBoundariesElement,
			popupsScrollableElement,
			disabled,
			dispatchAnalyticsEvent,
			pluginHooks,
			featureFlags,
		} = this.props;
		const maxContentSizeReached = Boolean(maxContentSize?.maxContentSizeReached);

		const editorViewMode = useSharedPluginStateWithSelector(
			this.props.editorAPI,
			['editorViewMode'],
			(states) => states?.editorViewModeState?.mode,
		);

		return (
			<WithFlash animate={maxContentSizeReached}>
				<ChromelessEditorContainer
					maxHeight={maxHeight}
					minHeight={minHeight}
					containerRef={(ref: HTMLElement | null) => (this.containerElement = ref)}
				>
					<EditorContentContainer
						// eslint-disable-next-line @atlaskit/ui-styling-standard/no-classname-prop
						className="ak-editor-content-area"
						featureFlags={featureFlags}
						viewMode={editorViewMode}
						appearance={this.appearance}
					>
						{customContentComponents && 'before' in customContentComponents
							? customContentComponents.before
							: customContentComponents}
						<PluginSlot
							editorView={editorView}
							editorActions={editorActions}
							eventDispatcher={eventDispatcher}
							providerFactory={providerFactory}
							appearance={this.appearance}
							items={contentComponents}
							popupsMountPoint={popupsMountPoint}
							popupsBoundariesElement={popupsBoundariesElement}
							popupsScrollableElement={popupsScrollableElement}
							containerElement={this.containerElement}
							disabled={!!disabled}
							dispatchAnalyticsEvent={dispatchAnalyticsEvent}
							wrapperElement={this.containerElement}
							pluginHooks={pluginHooks}
						/>
						{editorDOMElement}
						{customContentComponents && 'after' in customContentComponents
							? customContentComponents.after
							: null}
					</EditorContentContainer>
				</ChromelessEditorContainer>
			</WithFlash>
		);
	};

	render() {
		return (
			<RenderWithPluginState editorAPI={this.props.editorAPI} renderChrome={this.renderChrome} />
		);
	}
}

interface PluginStates {
	maxContentSize?: MaxContentSizePluginState;
}

interface RenderChromeProps extends Pick<AppearanceProps, 'editorAPI'> {
	renderChrome: (props: PluginStates) => React.ReactNode;
}

function RenderWithPluginState({ renderChrome, editorAPI }: RenderChromeProps) {
	const maxContentSizeReached = useSharedPluginStateWithSelector(
		editorAPI,
		['maxContentSize'],
		(states) => states?.maxContentSizeState?.maxContentSizeReached,
	);
	const maxContentSize =
		maxContentSizeReached === undefined ? undefined : { maxContentSizeReached };

	return <Fragment>{renderChrome({ maxContentSize })}</Fragment>;
}

interface ChromelessEditorContainerProps {
	children: React.ReactNode;
	containerRef?: (ref: HTMLElement | null) => void;
	maxHeight?: number;
	minHeight: number;
}

export function ChromelessEditorContainer({
	maxHeight,
	minHeight,
	children,
	containerRef,
}: ChromelessEditorContainerProps) {
	return (
		<div
			css={[
				chromelessEditorStylesNew,
				scrollbarStylesNew,
				fg('platform_fix_extra_space_last_line_comment_editor') && extraSpaceLastLineFix,
			]}
			style={{
				maxHeight: maxHeight ? `${maxHeight}px` : undefined,
				minHeight: `${minHeight}px`,
			}}
			data-testid="chromeless-editor"
			id="chromeless-editor"
			ref={containerRef}
		>
			{children}
		</div>
	);
}
