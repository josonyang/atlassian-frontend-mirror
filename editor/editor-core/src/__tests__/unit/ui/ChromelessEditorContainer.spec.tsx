import React from 'react';

import { render } from '@testing-library/react';

import { ChromelessEditorContainer } from '../../../ui/Appearance/Chromeless';

// this test is mainly to check if dev only updated legacy styles, and forget to update in the new styles
// if the new styles are not updated, we will see a difference in the snapshot
// and dev should update the new styles, and then update the snapshot
describe('ChromelessEditorContainer', () => {
	it('should have correct styles for new container', () => {
		const { container } = render(
			<ChromelessEditorContainer minHeight={100} maxHeight={200}>
				<div>child</div>
			</ChromelessEditorContainer>,
		);
		expect(container).toMatchSnapshot();
	});

	it('should have correct styles for new container with no max height', () => {
		const { container } = render(
			<ChromelessEditorContainer minHeight={100}>
				<div>child</div>
			</ChromelessEditorContainer>,
		);
		expect(container).toMatchSnapshot();
	});
});
