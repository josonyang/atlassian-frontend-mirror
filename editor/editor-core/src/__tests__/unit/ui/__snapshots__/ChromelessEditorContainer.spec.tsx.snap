// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromelessEditorContainer should have correct styles for new container 1`] = `
.emotion-0 {
  line-height: 20px;
  height: auto;
  overflow-x: hidden;
  overflow-y: auto;
  max-width: inherit;
  box-sizing: border-box;
  word-wrap: break-word;
  -ms-overflow-style: -ms-autohiding-scrollbar;
}

.emotion-0 div>.ProseMirror {
  outline: none;
  white-space: pre-wrap;
  padding: 0;
  margin: 0;
}

.emotion-0 div>.ProseMirror>:last-child {
  padding-bottom: var(--ds-space-100, 0.5em);
}

.emotion-0::-webkit-scrollbar-corner {
  display: none;
}

.emotion-0::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-subtle, #00000000);
}

.emotion-0:hover::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-bold, #44546F);
  border-radius: var(--ds-radius-large, 8px);
}

.emotion-0::-webkit-scrollbar-thumb:hover {
  background-color: var(--ds-background-neutral-bold-hovered, #2C3E5D);
}

<div>
  <div
    class="emotion-0"
    data-testid="chromeless-editor"
    id="chromeless-editor"
    style="max-height: 200px; min-height: 100px;"
  >
    <div>
      child
    </div>
  </div>
</div>
`;

exports[`ChromelessEditorContainer should have correct styles for new container with no max height 1`] = `
.emotion-0 {
  line-height: 20px;
  height: auto;
  overflow-x: hidden;
  overflow-y: auto;
  max-width: inherit;
  box-sizing: border-box;
  word-wrap: break-word;
  -ms-overflow-style: -ms-autohiding-scrollbar;
}

.emotion-0 div>.ProseMirror {
  outline: none;
  white-space: pre-wrap;
  padding: 0;
  margin: 0;
}

.emotion-0 div>.ProseMirror>:last-child {
  padding-bottom: var(--ds-space-100, 0.5em);
}

.emotion-0::-webkit-scrollbar-corner {
  display: none;
}

.emotion-0::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-subtle, #00000000);
}

.emotion-0:hover::-webkit-scrollbar-thumb {
  background-color: var(--ds-background-neutral-bold, #44546F);
  border-radius: var(--ds-radius-large, 8px);
}

.emotion-0::-webkit-scrollbar-thumb:hover {
  background-color: var(--ds-background-neutral-bold-hovered, #2C3E5D);
}

<div>
  <div
    class="emotion-0"
    data-testid="chromeless-editor"
    id="chromeless-editor"
    style="min-height: 100px;"
  >
    <div>
      child
    </div>
  </div>
</div>
`;
